pkg/oc/cli/cmd/push/binary.go

```diff
@@ -0,0 +1,363 @@
+		or as deployed applications. It accepts a zip or tar.gz file that will become a new 
+		image in an image stream. You may also specify image metadata like the entrypoint, 
+		environment variables, or labels to create a runnable image.
+
+		Experimental: This command is under active development and may change without notice.`)
+
+	pushBinaryExample = templates.Examples(`
+		# Upload myblob.zip as a new image in the app:binary image stream tag
+	  $ %[1]s binary --to=app:binary myblob.zip
+
+	  # Force upload in case another user is also uploading at the same time.
+	  $ %[1]s binary --to=app:binary myblob.zip --force`)
+)
+
+type PushBinaryOptions struct {
+	Out, ErrOut io.Writer
+	In          io.Reader
+	Output      string
+	PrintObject func([]*resource.Info) error
+	Mapper      *resource.Mapper
+
+	Path      string
+	BaseImage string
+	To        string
+	Namespace string
+	Force     bool
+
+	Client clientset.ImageStreamsGetter
+}
+
+// NewCmdPushBinary creates a new binary image in an image stream
+func NewCmdPushBinary(fullName string, f *clientcmd.Factory, in io.Reader, out, errout io.Writer) *cobra.Command {
+	options := &PushBinaryOptions{
+		Out:    out,
+		ErrOut: errout,
+		In:     in,
+	}
+	cmd := &cobra.Command{
+		Use:     "binary --to=IMAGESTREAMTAG [ARCHIVE|-]",
+		Short:   "Upload a zip or tar.gz file as an image",
+		Long:    pushBinaryLong,
+		Example: fmt.Sprintf(pushBinaryExample, fullName),
+		Run: func(cmd *cobra.Command, args []string) {
+			kcmdutil.CheckErr(options.Complete(f, cmd, args))
+			kcmdutil.CheckErr(options.Validate())
+			kcmdutil.CheckErr(options.Run())
+		},
+	}
+	cmd.MarkFlagRequired("filename")
+
+	cmd.Flags().StringVar(&options.BaseImage, "from", options.BaseImage, "An optional image stream tag or image to use as the base.")
+	cmd.Flags().StringVar(&options.To, "to", options.To, "An image stream tag to push to. Can be fully specified or relative to the current project.")
+	cmd.Flags().BoolVar(&options.Force, "force", options.Force, "If true, cancel any other pending uploads to this tag.")
+
+	kcmdutil.AddPrinterFlags(cmd)
+
+	return cmd
+}
+
+func (o *PushBinaryOptions) Complete(f *clientcmd.Factory, cmd *cobra.Command, args []string) error {
+	if len(args) != 1 {
+		return fmt.Errorf("you must specify exactly one argument - a URL, a path to a zip or tar.gz file, or '-' to use STDIN")
+	}
+	o.Path = args[0]
+
+	ns, _, err := f.DefaultNamespace()
+	if err != nil {
+		return err
+	}
+	o.Namespace = ns
+
+	config, err := f.ClientConfig()
+	if err != nil {
+		return err
+	}
+	o.Client, err = clientset.NewForConfig(config)
+
+	o.Output = kcmdutil.GetFlagString(cmd, "output")
+	o.PrintObject = func(infos []*resource.Info) error {
+		return f.PrintResourceInfos(cmd, false, infos, o.Out)
+	}
+	o.Mapper = clientcmd.ResourceMapper(f)
+
+	return err
+}
+
+func (o *PushBinaryOptions) Validate() error {
+	if len(o.To) == 0 {
+		return fmt.Errorf("--to=IMAGESTREAMTAG is a required argument")
```

Comment:
@dmage @smarterclayton It would be fun to add support for manifest lists here :)
I know that now we do not support them, but still.

Reply:
Yeah, we're going to have to deal with this sooner or later.  I may end up removing this command from this PR (so just the API) so we can discuss how we want to expose it separately.  The API is fairly self contained, while the commands get a lot into "what we want to do with this feature".  

If manifest lists can be updated later, then we have more flexibility to alter how copy works when used against a manifest list vs a single item.